src/Nerdbank.MessagePack/Converters/HardwareAccelerated.cs

```diff
@@ -0,0 +1,2634 @@
+/// </summary>
+internal static class HardwareAccelerated
+{
+	private enum SpanConstructorKind
+	{
+		Array,
+		List,
+		ReadOnlyMemory,
+		Memory,
+	}
+
+	/// <summary>
+	/// Creates a hardware-accelerated converter if one is available for the given enumerable and element type.
+	/// </summary>
+	/// <typeparam name="TEnumerable">The type of enumerable.</typeparam>
+	/// <typeparam name="TElement">The type of element.</typeparam>
+	/// <param name="converter">Receives the hardware-accelerated converter if one is available.</param>
+	/// <returns>A value indicating whether a converter is available.</returns>
+	internal static bool TryGetConverter<TEnumerable, TElement>([NotNullWhen(true)] out MessagePackConverter<TEnumerable>? converter)
+	{
+		Type enumerableType = typeof(TEnumerable);
+		SpanConstructorKind spanConstructorKind;
+		if (enumerableType == typeof(TElement[]))
+		{
+			spanConstructorKind = SpanConstructorKind.Array;
+		}
+		else if (enumerableType == typeof(List<TElement>))
+		{
+			spanConstructorKind = SpanConstructorKind.List;
+		}
+		else if (enumerableType == typeof(ReadOnlyMemory<TElement>))
+		{
+			spanConstructorKind = SpanConstructorKind.ReadOnlyMemory;
+		}
+		else if (enumerableType == typeof(Memory<TElement>))
+		{
+			spanConstructorKind = SpanConstructorKind.Memory;
+		}
+		else
+		{
+			goto FAIL;
+		}
+
+		if (typeof(TElement) == typeof(bool))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, bool>(spanConstructorKind);
+			return true;
+		}
+		else if (typeof(TElement) == typeof(sbyte))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, sbyte>(spanConstructorKind);
+			return true;
+		}
+		else if (typeof(TElement) == typeof(short))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, short>(spanConstructorKind);
+			return true;
+		}
+		else if (typeof(TElement) == typeof(int))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, int>(spanConstructorKind);
+			return true;
+		}
+		else if (typeof(TElement) == typeof(long))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, long>(spanConstructorKind);
+			return true;
+		}
+		else if (typeof(TElement) == typeof(ushort))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, ushort>(spanConstructorKind);
+			return true;
+		}
+		else if (typeof(TElement) == typeof(uint))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, uint>(spanConstructorKind);
+			return true;
+		}
+		else if (typeof(TElement) == typeof(ulong))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, ulong>(spanConstructorKind);
+			return true;
+		}
+		else if (typeof(TElement) == typeof(float))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, float>(spanConstructorKind);
+			return true;
+		}
+		else if (typeof(TElement) == typeof(double))
+		{
+			converter = new PrimitiveArrayConverter<TEnumerable, double>(spanConstructorKind);
+			return true;
+		}
+
+FAIL:
```

Comment:
This label is used in just one place. Why not return from there directly?

Reply:
I do not like duplicated code but I like early-return.
I tend to use goto statement in order to go to default failure code path.
Ok, I remove this goto statement.